tests/UnitTests/Foundation/DispatchTests.mm

```diff
@@ -0,0 +1,122 @@
+
+	dispatch_group_async(group, queue, ^{
+		EXPECT_EQ(dispatch_get_specific(key), context);
+        dispatch_queue_set_specific(queue, key, context2, NULL);
+    });
+	dispatch_group_wait(group, DISPATCH_TIME_FOREVER);
+    
+	EXPECT_EQ(dispatch_queue_get_specific(queue, key), context2);
+    
+    dispatch_group_async(group, queue, ^{
+		EXPECT_EQ(dispatch_get_specific(key), context2);
+    });
+	dispatch_group_wait(group, DISPATCH_TIME_FOREVER);
+}
+
+TEST(DispatchTests, MultipleSpecificKeys) {
+    dispatch_queue_t queue = dispatch_queue_create("queue", NULL);
+	dispatch_group_t group = dispatch_group_create();
+
+    dispatch_queue_set_specific(queue, key, context, NULL);
+	dispatch_queue_set_specific(queue, key2, context2, NULL);
+    
+    dispatch_group_async(group, queue, ^{
+		EXPECT_EQ(dispatch_get_specific(key), context);
+		EXPECT_EQ(dispatch_get_specific(key2), context2);
+        dispatch_queue_set_specific(queue, key, context2, NULL);
+		dispatch_queue_set_specific(queue, key2, context, NULL);
+    });
+	dispatch_group_wait(group, DISPATCH_TIME_FOREVER);
+    
+	EXPECT_EQ(dispatch_queue_get_specific(queue, key), context2);
+	EXPECT_EQ(dispatch_queue_get_specific(queue, key2), context);
+
+	dispatch_queue_set_specific(queue, key3, context3, NULL);
+    
+    dispatch_group_async(group, queue, ^{
+		EXPECT_EQ(dispatch_get_specific(key), context2);
+		EXPECT_EQ(dispatch_get_specific(key2), context);
+		EXPECT_EQ(dispatch_get_specific(key3), context3);
+    });
+	dispatch_group_wait(group, DISPATCH_TIME_FOREVER);
+}
+
+TEST(DispatchTests, SpecificDestructor) {
+	dispatch_queue_t queue = dispatch_queue_create("queue", NULL);
+	dispatch_group_t group = dispatch_group_create();
+
+    dispatch_queue_set_specific(queue, key, context, destructor);
+
+	EXPECT_EQ(destructorCount, 0);
+    
+    dispatch_group_async(group, queue, ^{
+		EXPECT_EQ(dispatch_get_specific(key), context);
+        dispatch_queue_set_specific(queue, key, context2, destructor);
+		EXPECT_EQ(destructorCount, 1);
+    });
+	dispatch_group_wait(group, DISPATCH_TIME_FOREVER);
+    
+	EXPECT_EQ(dispatch_queue_get_specific(queue, key), context2);
+	dispatch_queue_set_specific(queue, key, context, destructor);
+	EXPECT_EQ(destructorCount, 2);
+    
+    dispatch_group_async(group, queue, ^{
+		EXPECT_EQ(dispatch_get_specific(key), context);
+		EXPECT_EQ(destructorCount, 2);
+    });
+	dispatch_group_wait(group, DISPATCH_TIME_FOREVER);
+}
+
+TEST(DispatchTests, RemoveSpecific) {
+	dispatch_queue_t queue = dispatch_queue_create("queue", NULL);
+	dispatch_group_t group = dispatch_group_create();
+
+    dispatch_queue_set_specific(queue, key, context, NULL);
+    
+    dispatch_group_async(group, queue, ^{
+		EXPECT_EQ(dispatch_get_specific(key), context);
+        dispatch_queue_set_specific(queue, key, NULL, NULL);
+    });
+	dispatch_group_wait(group, DISPATCH_TIME_FOREVER);
+
+	if(dispatch_queue_get_specific(queue, key)) {
+		EXPECT_EQ(1, 0);
+	}
+}
```

Comment:
I'd like to see a test that involves multiple queues, just to make sure that we never get cross-contamination.